### PR TITLE
Implement Hbase2.x replication endpoint 

### PR DESCRIPTION
**Pull request by sushanb**
- [x] Removes WAL.Entry object from hbase-replication-libary
- [x] Separates 1.x and 2x replication libaries
- [x] Introduces BigtableWALEntry

### CODE REVIEW

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static java.util.stream.Collectors.groupingBy;
+
+
+public class HbaseToCloudBigtableReplicationEndpoint extends BaseReplicationEndpoint {
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java:16]: BaseReplicationEndpoint was marked as Private when they broke compatiblity on hbase 2.0 

" It undoes the LimitedPrivate on BaseReplicationEndpoint while keeping it for the RE Interface. It means consumers will have to copy/paste the AbstractService-based BRE into their own codebase also supplying their  own Guava;"
https://issues.apache.org/jira/browse/HBASE-15982

You should implement the ReplicationEndpoint interface and implement/copy-paste the guava service as they suggested in the jira.


BaseReplicationEndpoint is `Private`
https://github.com/apache/hbase/blob/9f7e856a5f84a6f52d0197a03b7a56e1a3f0446c/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/BaseReplicationEndpoint.java#L36

ReplicationEndpoint is `LimitedPrivate` and we should use it.
https://github.com/apache/hbase/blob/9f7e856a5f84a6f52d0197a03b7a56e1a3f0446c/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/ReplicationEndpoint.java#L50

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public void start() {
+        startAsync();
+
+    }
+
+    @Override
+    public void stop() {
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java:46]: Yeah, we need to implement these methods. you can't use BaseReplicationEndpoint

File: bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(JUnit4.class)
+public class HbaseToCloudBigtableReplicationEndpointTest {
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java:27]: Is this an exact copy of 1.x tests with a 2.1 utility? 

We should also add a test with SERIAL_REPLICATION is enabled.

File: bigtable-hbase-replication/pom.xml [hunk 4]
```diff
@@ -201,12 +201,6 @@
       <version>3.8.0</version>
       <scope>test</scope>
     </dependency>
-    <!--    <dependency>-->
```
Comment (vermas2012) [inline bigtable-hbase-replication/pom.xml:204]: We should also remove the dependency on the HBase-server, if not in this PR then in the next one.

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java [hunk 5]
```diff
@@ -169,15 +171,15 @@ public synchronized void start(Configuration configuration, MetricsSource metric
...
+  public void stop() {
 
     LOG.error("Stopping replication to CBT for this EndPoint. ",
         new RuntimeException("Dummy exception for stacktrace"));
 
     // Connection is shared by all the instances of this class, close it only if no one is using it.
     // Closing the connection is required as it owns the underlying gRpc connections and gRpc does
     // not like JVM shutting without closing the gRpc connections.
-    synchronized (HbaseToCloudBigtableReplicationEndpoint.class) {
+    synchronized (this) {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:182]: This was synchronized on the class as its locking static variables. Making it lock on the object may lead to race conditions.

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    private ArrayList<Cell> cells;
+    private String tableName;
+
+    public BigtableWALEntry(long timeStamp, ArrayList<Cell> cells, String tableName) {
+        this.timeStamp = timeStamp;
+        this.cells = cells;
+        this.tableName = tableName;
+    }
+
+    public ArrayList<Cell> getWalEdit() {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java:20]: How about renaming it to getCells()?

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.google.cloud.bigtable.hbase.replication.adapters;
+
+/*
+HbaseBigtableWALEntryAdaptor provides BigtableWALEntry adaptor for any WAL entry
+ */
+public interface HbaseBigtableWALEntryAdaptor {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java:6]: This adaptor interface is not really "adapting". To define the adaptor correctly ( newType adapt(OldType old)) you need to depend on the "WAL.Entry" which defeats the purpose of creating BigtableWALEntry. 

This package should use BigtableWALEntry everywhere, it is the callers responsibility to figure out how to create it. We can delete this file here.

**Review by vermas2012 — CHANGES_REQUESTED** [review]

File: bigtable-hbase-1.x-replication/pom.xml [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+            <version>0.125.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>3.8.0</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (sushanb) [reply bigtable-hbase-1.x-replication/pom.xml]: Done.

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java (hunk 7 above)
Comment (sushanb) [reply bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java]: Deleted the file

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java (hunk 6 above)
Comment (sushanb) [reply bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java]: Done.

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java (hunk 5 above)
Comment (sushanb) [reply bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java]: synchronized on the class

File: bigtable-hbase-replication/pom.xml (hunk 4 above)
Comment (sushanb) [reply bigtable-hbase-replication/pom.xml:204]: Will do it in next PR. remove ReplicationSource

File: bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java (hunk 3 above)
Comment (sushanb) [reply bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java]: it is not exact copy as some methods are deprecated. Will add a test with SERIAL_REPLICATION

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java (hunk 2 above)
Comment (sushanb) [reply bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java]: Will do it in next PR.

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java (hunk 7 above)
Comment (sushanb) [reply bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/HbaseBigtableWALEntryAdaptor.java]: upcoming PR.

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/BigtableWALEntryImpl.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import com.google.cloud.bigtable.hbase.replication.adapters.HbaseBigtableWALEntryAdaptor;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.wal.WAL;
+
+import java.util.ArrayList;
+
+/*
+BigtableWALEntryImpl provides BigWALEntry class from hbase1x. WAL.Entry
+ */
+public class BigtableWALEntryImpl implements HbaseBigtableWALEntryAdaptor {
```
[the remark this answers is not in the document]
Comment (sushanb) [reply bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/BigtableWALEntryImpl.java]: Done. Removed adaptor

File: bigtable-hbase-2.x-replication/pom.xml [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            <version>0.125.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>3.8.0</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (sushanb) [reply bigtable-hbase-2.x-replication/pom.xml]: Done.

File: bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 11]
```diff
@@ -13,13 +13,20 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.google.cloud.bigtable.hbase.replication;
+package com.google.cloud.bigtable.hbase1_x.replication;
```
[the remark this answers is not in the document]
Comment (sushanb) [reply bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpoint.java]: same package name across all versions.

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java (hunk 1 above)
Comment (sushanb) [reply bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java]: upcoming PR.

File: bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/BigtableWALEntryImpl.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.google.cloud.bigtable.hbase.replication.adapters.BigtableWALEntry;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.wal.WAL;
+
+import java.util.ArrayList;
+/*
+BigtableWALEntryImpl provides BigWALEntry class from hbase1x. WAL.Entry
+ */
+public class BigtableWALEntryImpl  {
```
Comment (vermas2012) [inline bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/BigtableWALEntryImpl.java:11]: This class can be in the core package. You don't need to depend on the WAL. you can create a constructor  liek

```
BigtableWalEntry(String tableName, List<Cell> cells, long writeTimeInMillis)
```

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/BigtableWALEntryImpl.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import com.google.cloud.bigtable.hbase.replication.adapters.BigtableWALEntry;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.wal.WAL;
+
+import java.util.ArrayList;
+
+/*
+BigtableWALEntryImpl provides BigWALEntry class from hbase1x. WAL.Entry
+ */
+public class BigtableWALEntryImpl  {
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/BigtableWALEntryImpl.java:12]: same, you don't need this class.

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public UUID getPeerUUID() {
+        return cloudBigtableReplicator.getPeerUUID(); }
+
+    @Override
+    public boolean replicate(ReplicateContext replicateContext) {
+        Map<String, List<BigtableWALEntry>> walEntriesByTable = new HashMap<>();
+        for (WAL.Entry wal: replicateContext.getEntries()) {
+            String tableName = wal.getKey().getTableName().getNameAsString();
+            BigtableWALEntryImpl bigtableWALEntryImpl = new BigtableWALEntryImpl(wal);
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:30]: Here, you can just call the core BigtableEntry constrcutor, you won't need the BigtableEntryImpl class.

File: bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.google.cloud.bigtable.hbase.replication.adapters.BigtableWALEntry;
+import org.apache.hadoop.hbase.replication.BaseReplicationEndpoint;
+import org.apache.hadoop.hbase.wal.WAL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+// TODO(remove BaseReplicationEndpoint extension).
+public class HbaseToCloudBigtableReplicationEndpoint extends BaseReplicationEndpoint {
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:11]: I thought we were moving to ReplicationEndpoint interface?

File: bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 16]
```diff
@@ -0,0 +1,321 @@
+//package com.google.cloud.bigtable.hbase.replication;
```
Comment (vermas2012) [inline bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:1]: Why is this disabled?

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java [hunk 17]
```diff
@@ -195,34 +195,21 @@ public UUID getPeerUUID() {
     return UUID.nameUUIDFromBytes("Cloud-bigtable".getBytes(StandardCharsets.UTF_8));
   }
 
-  public boolean replicate(List<WAL.Entry> walsToReplicate) {
+  public boolean replicate(Map<String, List<BigtableWALEntry>> walEntriesByTable) {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:198]: Interesting, how are you making it work? The BigtableWALEntryImpl classes don't extend/implement the BigtableWALEntry, how does it work?

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.ArrayList;
+/**
+ * BigtableWALEntry abstracts minimal functionality from WAL.Entry required for this replication library.
+ */
+public class BigtableWALEntry {
+    private long timeStamp;
+    private ArrayList<Cell> cells;
+    private String tableName;
+
+    public BigtableWALEntry(long timeStamp, ArrayList<Cell> cells, String tableName) {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java:15]: Exactly ! call this constructor from the 1.x and 2.x endpoints and you dont' need the IMPL files.

**Review by vermas2012 — CHANGES_REQUESTED** [review]

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java [hunk 19]
```diff
@@ -10,12 +10,10 @@
 public class BigtableWALEntry {
     private long timeStamp;
     private ArrayList<Cell> cells;
-    private String tableName;
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java]: Why did we remove the table name? Do we have access to the table name in the IncompatibleMutationAdapter? It is important there. If we have to query CBT for the incompatible mutation, we need the table name.

**Pull request merged**